build/utf8/valid_asm.go

```diff
@@ -0,0 +1,367 @@
+	VPSHUFB(lowPrev, nibble2Y, lowPrev)
+	VPAND(lowPrev, highPrev, highPrev)
+
+	Comment("Check errors on the high nibble on the current byte")
+	highCurr := highNibbles(currentBlockY, nibbleMaskY)
+	VPSHUFB(highCurr, nibble3Y, highCurr)
+	VPAND(highCurr, highPrev, highPrev)
+
+	Comment("Find 2 bytes continuations")
+	off2 := pushLast2BytesFromAToFrontOfB(previousBlockY, currentBlockY)
+	VPSUBUSB(continuation3BytesY, off2, off2)
+
+	Comment("Find 3 bytes continuations")
+	off3 := pushLast3BytesFromAToFrontOfB(previousBlockY, currentBlockY)
+	VPSUBUSB(continuation4BytesY, off3, off3)
+
+	Comment("Combine them to have all continuations")
+	continuationBitsY := YMM()
+	VPOR(off2, off3, continuationBitsY)
+
+	Comment("Perform a byte-sized signed comparison with zero to turn any non-zero bytes into 0xFF.")
+	tmpY := zeroOutVector(YMM())
+	VPCMPGTB(tmpY, continuationBitsY, continuationBitsY)
+
+	Comment("Find bytes that are continuations by looking at their most significant bit.")
+	msbMask := ConstArray64("msb_mask",
+		0x8080808080808080,
+		0x8080808080808080,
+		0x8080808080808080,
+		0x8080808080808080,
+	)
+
+	msbMaskY := YMM()
+	VMOVDQU(msbMask, msbMaskY)
+	VPAND(msbMaskY, continuationBitsY, continuationBitsY)
+
+	Comment("Find mismatches between expected and actual continuation bytes")
+	VPXOR(continuationBitsY, highPrev, continuationBitsY)
+
+	Comment("Store result in sticky error")
+	VPOR(errorY, continuationBitsY, errorY)
+
+	Comment("Prepare for next iteration")
+	VPSUBUSB(incompleteMaskY, currentBlockY, incompletePreviousBlockY)
+	VMOVDQU(currentBlockY, previousBlockY)
+
+	Comment("End of loop")
+	JMP(LabelRef("check_input"))
+
+	Label("end")
+
+	Comment("If the previous block was incomplete, this is an error.")
+	VPOR(incompletePreviousBlockY, errorY, errorY)
+
+	Comment("Return whether any error bit was set")
+	VPTEST(errorY, errorY)
+	SETEQ(ret.Addr)
+	RET()
+
+	Generate()
+}
+
+func pushLast2BytesFromAToFrontOfB(a, b VecVirtual) VecVirtual {
+	out := YMM()
+	VPERM2I128(Imm(3), a, b, out)
+	VPALIGNR(Imm(14), out, b, out)
+	return out
+}
+
+func pushLast3BytesFromAToFrontOfB(a, b VecVirtual) VecVirtual {
+	out := YMM()
+	VPERM2I128(Imm(3), a, b, out)
+	VPALIGNR(Imm(13), out, b, out)
+	return out
+}
+
+func pushLastByteFromAToFrontOfB(a, b VecVirtual) VecVirtual {
+	out := YMM()
+	VPERM2I128(Imm(3), a, b, out)
+	VPALIGNR(Imm(15), out, b, out)
+	return out
+}
+
+func lowNibbles(a VecVirtual, nibbleMask VecVirtual) VecVirtual {
+	out := YMM()
+	VPAND(a, nibbleMask, out)
+	return out
+}
+
+func highNibbles(a VecVirtual, nibbleMask VecVirtual) VecVirtual {
+	out := YMM()
+	VPSRLW(Imm(4), a, out)
+	VPAND(out, nibbleMask, out)
+	return out
+}
+
+func zeroOutVector(y VecVirtual) VecVirtual {
+	VXORPS(y, y, y)
+	return y
+}
+
+// Assumptions:
+// - len(dst) == 32 bytes
+// - 0 < len(src) < 32
+// TODO: try to use x86.bytes
+func copyN(dst Register, src Register, n Register) {
+	v := x86.VariableLengthBytes{
+		Unroll: 128,
```

Comment:
It seems we are using `copyN` on a scratch buffer of 64 bytes, does it make sense for the unroll loop to treat up to 128 bytes?

Reply:
Forgot to remove that line. In the current implementation I don't think it matters, because I don't call `v.Generate()`, so the `Unroll` field is never used.